phi/tools/googlecalender.py

```diff
@@ -0,0 +1,173 @@
+    )
+from typing import List, Optional
+
+SCOPES = ["https://www.googleapis.com/auth/calendar"]
+
+
+def authenticated(func):
+    """Decorator to ensure authentication before executing the method."""
+
+    @wraps(func)
+    def wrapper(self, *args, **kwargs):
+        # Ensure credentials are valid
+        if os.path.exists(self.token_path):
+            self.creds = Credentials.from_authorized_user_file(self.token_path, SCOPES)
+        if not self.creds or not self.creds.valid:
+            if self.creds and self.creds.expired and self.creds.refresh_token:
+                self.creds.refresh(Request())
+            else:
+                flow = InstalledAppFlow.from_client_secrets_file(self.creds_path, SCOPES)
+                self.creds = flow.run_local_server(port=0)
+                # Save the credentials for future use
+            with open(self.token_path, "w") as token:
+                token.write(self.creds.to_json())
+
+            # Initialize the Google Calendar service
+        try:
+            self.service = build("calendar", "v3", credentials=self.creds)
+        except HttpError as error:
+            logger.error(f"An error occurred while creating the service: {error}")
+            raise
+
+        # Ensure the service is available
+        if not self.service:
+            raise ValueError("Google Calendar service could not be initialized.")
+
+        return func(self, *args, **kwargs)
+
+    return wrapper
+
+
+class GoogleCalenderTools(Toolkit):
+    def __init__(self, credentials_path: Optional[str] = None, token_path: Optional[str] = None):
+        """
+        Google Calender Tool.
+
+        :param credentials_path: Path of the file credentials.json file which contains OAuth 2.0 Client ID. A client ID is used to identify a single app to Google's OAuth servers. If your app runs on multiple platforms, you must create a separate client ID for each platform. Refer doc https://developers.google.com/calendar/api/quickstart/python#authorize_credentials_for_a_desktop_application
+        :param token_path: Path of the file token.json which stores the user's access and refresh tokens, and is created automatically when the authorization flow completes for the first time.
+
+        """
+        super().__init__(name="google_calender_tools")
+
+        if not credentials_path:
+            logger.error(
+                "Google Calender Tool : Please Provide Valid Credentials Path , You can refer https://developers.google.com/calendar/api/quickstart/python#authorize_credentials_for_a_desktop_application to create your credentials"
+            )
+            raise ValueError("Credential path is required")
+        elif not os.path.exists(credentials_path):
+            logger.error(
+                "Google Calender Tool : Credential file Path is invalid , please provide the full path of the credentials json file"
+            )
+            raise ValueError("Credentials Path is invalid")
+
+        if not token_path:
```

Comment:
User authentication data is saved to a shared space, and everyone else will have access to someone else's calendar

Reply:
Thank you for raising this concern! I’d like to clarify that in current setup I'm assuming each user operates within their own deployment, and the token.json file is stored locally within the deployment environment.  I’m not entirely sure about the specifics of how deployments and storage are handled in the current PhiData app setup. I would love to get input from the maintainers here to ensure we’re following best practices for securely handling credentials.

If there’s a recommended approach for isolating tokens or managing user-specific authentication, I’m happy to implement that.

Reply:
Please check the use of user_id examples. If the token will be stored as “{user_id}_token”, this will be enough for the first version.

The best way might be to use user storage, but that's a bit more complex.